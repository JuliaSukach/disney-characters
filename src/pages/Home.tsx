import React, { useState, useEffect } from 'react'
import { Link } from 'react-router-dom'
import useDebounce from '../hooks/useDebounce'
import usePagination from '../hooks/usePagination'
import useUrlParams from '../hooks/useUrlParams'
import useCharacterSearch from '../hooks/useCharacterSearch'
import PaginationButton from '../components/PaginationButton'
import Spinner from '../components/Spinner'
import ErrorMessage from '../components/ErrorMessage'
import { createPaginationRange } from '../utils/pagination'
import '../css/spinner.css'
import '../css/home.css'

const Home: React.FC = () => {
    // Custom hook for URL parameters
    const { initialSearchTerm, initialPage, initialPageSize, updateUrlParams } = useUrlParams()

    // State Management
    const [searchTerm, setSearchTerm] = useState<string>(initialSearchTerm)
    const [totalPages, setTotalPages] = useState<number>(1)

    // Use custom pagination hook
    const { page, setPage } = usePagination(initialPage)

    // Debounce the search term so API calls are not made on every keystroke
    const debouncedSearchTerm = useDebounce(searchTerm, 500)

    // Custom hook for character search logic
    const { results, loading, error, fetchCharactersForSearchTerm, resetResults, setError } = useCharacterSearch()


    // Fetch characters whenever debounced search term or page changes
    useEffect(() => {
        if (debouncedSearchTerm.length < 1) {
            resetResults()
            setTotalPages(1)
            setPage(1)
            setError('')
            return
        }

        // Function to fetch characters
        const getCharacters = async () => {
            resetResults()
            setError('')
            try {
                const { totalPages: fetchedTotalPages } = await fetchCharactersForSearchTerm(debouncedSearchTerm, page, initialPageSize)
                // Set total pages
                setTotalPages(fetchedTotalPages)
            } catch (err) {
                console.error('Failed to fetch characters:', err)
            }
        }

        getCharacters()
    }, [debouncedSearchTerm, page])

    // Use another useEffect for updating URL params when relevant state changes
    useEffect(() => {
        if (debouncedSearchTerm.length < 1) {
            updateUrlParams({ q: '', page: '', pageSize: '' })
        } else {
            updateUrlParams({ q: debouncedSearchTerm, page: page.toString(), pageSize: initialPageSize.toString() })
        }
    }, [debouncedSearchTerm, page, updateUrlParams])

    // Handle input changes and update URL query
    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = e.target.value
        setSearchTerm(value)
        setPage(1)

        if (value === '') {
            setSearchTerm('')
        }
    }

    // Handle page number click
    const handlePageClick = (pageNumber: number) => {
        if (pageNumber !== page) {
            setPage(pageNumber)
        }
    }

    return (
        <div className='home-container'>
            <h1>Search Disney Characters</h1>
            <input
                className='search-input'
                type='text'
                value={searchTerm}
                onChange={handleInputChange}
                placeholder='Type a character name'
                aria-label='Search for Disney characters'
            />
            {loading && <Spinner />}
            {error && <ErrorMessage message={error} />}
            {!loading &&(
                <ul className='search-results'>
                {results.map((character) => (
                    <li key={character._id} id={`character-${character._id}`} className='search-result-item' tabIndex={0}>
                        <Link to={`/details/${character._id}`}>{character.name}</Link>
                    </li>
                ))}
            </ul>
            )}
            {totalPages > 1 && !loading && results.length > 0 && (
                <div className='pagination'>
                    {createPaginationRange(page, totalPages).map((pageItem: number | string) => (
                        <PaginationButton
                            key={pageItem}
                            pageNumber={pageItem}
                            isActive={page === pageItem}
                            onClick={() => typeof pageItem === 'number' && handlePageClick(pageItem)}
                            isDisabled={pageItem === '...'}
                        />
                    ))}
                </div>
            )}
        </div>
    )
}

export default Home