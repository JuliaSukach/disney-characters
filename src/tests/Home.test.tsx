import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import axios from 'axios'
import Home from '../pages/Home'

// Mock axios to avoid real API requests
jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('Home component', () => {
    beforeEach(() => {
        mockedAxios.get.mockClear() // Clear mock between tests
    })

    test('renders search input', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i)
        expect(inputElement).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot() // Snapshot of the initial render
    })

    test('updates search term on input change', () => {
        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i) as HTMLInputElement

        // Simulate user typing into the input
        fireEvent.change(inputElement, { target: { value: 'Mickey' } })

        expect(inputElement.value).toBe('Mickey') // Ensure input state updates
    })

    test('displays loading spinner when fetching', async () => {
        mockedAxios.get.mockResolvedValueOnce({ data: { data: [] } }) // Mock resolved API response

        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i)

        // Simulate input change to trigger fetching
        fireEvent.change(inputElement, { target: { value: 'Mickey' } })

        // Assert loading spinner is in the document
        await waitFor(() => {
            expect(screen.getByRole('status')).toBeInTheDocument()
        })

        // Snapshot of loading state
        await waitFor(() => {
            expect(screen.getByRole('status')).toMatchSnapshot()
        })
        
        // Wait for spinner to disappear
        await waitFor(() => {
            expect(screen.queryByRole('status')).not.toBeInTheDocument()
        })
    })

    test('displays search results after fetching data', async () => {
        const mockData = [
            { _id: '1', name: 'Mickey Mouse' },
            { _id: '2', name: 'Donald Duck' }
        ]
        mockedAxios.get.mockResolvedValueOnce({ data: { data: mockData, totalPages: 1 } })

        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i)

        fireEvent.change(inputElement, { target: { value: 'Mickey' } })

        // Wait for results to be displayed
        await waitFor(() => {
            expect(screen.getByText(/Mickey Mouse/i)).toBeInTheDocument()
            expect(screen.getByText(/Donald Duck/i)).toBeInTheDocument()
        })

        // Snapshot of the populated state
        expect(screen.getByText(/Mickey Mouse/i).parentElement).toMatchSnapshot()
    })

    test('shows error message on fetch error', async () => {
        mockedAxios.get.mockRejectedValueOnce(new Error('Network error. Please check your internet connection and try again.')) // Mock API error

        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i)

        // Simulate input change to trigger API call
        fireEvent.change(inputElement, { target: { value: 'Mickey' } })

        // Assert error message is shown
        await waitFor(() => {
            expect(screen.getByText(/Error fetching Disney characters. Please try again./i)).toBeInTheDocument()
        })

        // Snapshot of the error state
        expect(screen.getByText(/Error fetching Disney characters. Please try again./i)).toMatchSnapshot()
    })

    test('clears results when input is cleared', async () => {
        const mockData = [
            { _id: '1', name: 'Mickey Mouse' }
        ]
        mockedAxios.get.mockResolvedValueOnce({ data: { data: mockData, totalPages: 1 } })

        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )

        const inputElement = screen.getByPlaceholderText(/Type a character name/i)
        fireEvent.change(inputElement, { target: { value: 'Mickey' } })

        await waitFor(() => {
            expect(screen.getByText(/Mickey Mouse/i)).toBeInTheDocument()
        })

        // Clear the input
        fireEvent.change(inputElement, { target: { value: '' } })

        // Assert results are cleared
        await waitFor(() => {
            expect(screen.queryByText(/Mickey Mouse/i)).not.toBeInTheDocument()
        })
    })

    test('retains focus on input after typing', () => {
        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )

        const inputElement = screen.getByPlaceholderText(/Type a character name/i)
        inputElement.focus()
        expect(inputElement).toHaveFocus()

        fireEvent.change(inputElement, { target: { value: 'Donald' } })
        expect(inputElement).toHaveFocus() // Ensure input still has focus after change
    })

    test('displays no characters found message', async () => {
        mockedAxios.get.mockResolvedValueOnce({ data: { data: [], totalPages: 1 } })

        render(
            <MemoryRouter>
                <Home />
            </MemoryRouter>
        )
        const inputElement = screen.getByPlaceholderText(/Type a character name/i)

        // Simulate input change to trigger API call
        fireEvent.change(inputElement, { target: { value: 'Unknown' } })

        // Wait for the "No characters found." message to be displayed
        await waitFor(() => {
            expect(screen.getByText(/No characters found./i)).toBeInTheDocument()
        })

        // Snapshot of no results state
        expect(screen.getByText(/No characters found./i)).toMatchSnapshot()
    })
})