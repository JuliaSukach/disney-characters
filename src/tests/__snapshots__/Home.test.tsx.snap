// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component displays loading spinner when fetching 1`] = `
<div
  aria-label="Loading"
  class="spinner"
  role="status"
/>
`;

exports[`Home component displays no characters found message 1`] = `
<p
  class="error-message"
  role="alert"
  style="color: red;"
>
  No characters found.
</p>
`;

exports[`Home component displays search results after fetching data 1`] = `
<li
  class="search-result-item"
  id="character-1"
  tabindex="0"
>
  <a
    href="/details/1"
  >
    Mickey Mouse
  </a>
</li>
`;

exports[`Home component renders search input 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <h1>
      Search Disney Characters
    </h1>
    <input
      aria-label="Search for Disney characters"
      class="search-input"
      placeholder="Type a character name"
      type="text"
      value=""
    />
    <ul
      class="search-results"
    />
  </div>
</DocumentFragment>
`;

exports[`Home component shows error message on fetch error 1`] = `
<p
  class="error-message"
  role="alert"
  style="color: red;"
>
  Error fetching Disney characters. Please try again.
</p>
`;
